core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java

```diff
@@ -433,6 +433,33 @@ void clone(String srcTableName, String newTableName, boolean flush,
       Map<String,String> propertiesToSet, Set<String> propertiesToExclude) throws AccumuloException,
       AccumuloSecurityException, TableNotFoundException, TableExistsException;
 
+  /**
+   * Clone a table from an existing table. The cloned table will have the same data as the source
+   * table it was created from. After cloning, the two tables can mutate independently. Initially
+   * the cloned table should not use any extra space, however as the source table and cloned table
+   * major compact extra space will be used by the clone.
+   *
+   * Initially the cloned table is only readable and writable by the user who created it.
+   *
+   * @param srcTableName
+   *          the table to clone
+   * @param newTableName
+   *          the name of the clone
+   * @param flush
+   *          determines if memory is flushed in the source table before cloning.
+   * @param propertiesToSet
+   *          the sources tables properties are copied, this allows overriding of those properties
+   * @param propertiesToExclude
+   *          do not copy these properties from the source table, just revert to system defaults
+   * @param keepOffline
+   *          do not bring the table online after cloning
+   */
+
+  void clone(String srcTableName, String newTableName, boolean flush,
```

Comment:
Would be nice to create a simple class for clone configuration so in the future we do not need to add more overloaded methods.  Something like the following.

```java
/**
 * @since 2.1.0
 */
void clone(String srcTableName, String newTableName, CloneConfiguration config);
```

Could possibly have the following API for CloneConfiguration.

```java
/**
 * @since 2.1.0
 */
public class CloneConfiguration {
  Map<String,String> getPropertiesToSet();

  Set<String> getPropertiesToExclude();

  boolean flush();

  InitialTableState getInitialState();

  /**
   * @since 2.1.0
   */
  public static interface Builder {

    /**
     * props to set on the clone. If not set defaults to empty map.
     */
    Builder setProperties(Map<String,String> props);

    /**
     * Props to exclude form the source table.  If not set defaults to empty set.
     * 
     */
    Builder excludeProperties(Set<String> props);

    /**
     * The initial table state for the clone. If not set defaults to ONLINE.
     */
    Builder setInitialTableState(InitialTableState init);

    /**
     * Determines if source tables in memory data is flushed before cloning. Unflushed data will not be in clone.  If not set defaults to true.
     */
    Builder setFlush(boolean flush);

    // builds an immutable CloneConfiguration
    CloneConfiguration build();
  }

  public static Builder builder() {
    // TODO return a new builder obj that is not visible in the public API.. see ScanDirectivesImpl
    // for an example of simple way to implement
    return null;
  }
}
```

Reply:
I like it, but perhaps I do this only in the 2.1 line?

Reply:
I agree.  There shouldn't be anything in this change that can't go into 1.10 (now that it is java 1.8).  It should then be fairly easy to merge forward and no need for #1476

Reply:
ok, adding into the 1.10 line.  Change will be coming soon.

Reply:
I have implemented a CloneConfiguration and CloneConfigurationImpl.  However I left the impl public to enable creating a new instance in the CloneConfiguration.builder() implementation.  Looking at your ScanDirectives.builder(), it appears to always return the same instance of a builder and hence the configuration object.  Hence that is really not usable across threads.  Was that intentional?

Reply:
FYI ScanDirectives were added in https://github.com/apache/accumulo/pull/1440

Reply:
@ivakegg I think I figured out what is going on with the ScanDirectives.builder()... the setters in DefaultScanDirectives will return a new object of the parent class.  

```return new ScanDirectivesImpl().setExecutorName(name);```

So the builder will always return ```DefaultScanDirectives``` first with the ```builder()``` method and then only if something needs to be set return a new ```ScanDirectivesImpl``` object.

Reply:
Ah, I see it now.  So should I change to the same pattern for the builder()?  I guess I would at least be able to make the CloneConfigurationImpl package private if I did but I am not sure it is worth the extra class.

Reply:
The pattern I chose for ScanDirectives builder is complex because I thought it would be called very frequently and wanted it to be efficient.  For CloneConfiguration builder, I don't think efficiency would be a concern so personally I would go with the simplest implementation.

Reply:
OK, then I will leave it as is.  Thanks @keith-turner.